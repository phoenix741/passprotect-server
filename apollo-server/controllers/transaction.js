import { checkPermission } from '../utils/authentification'
import { getTransactions } from '../services/transaction'
import { pubsub, TRANSACTION_ADDED_TOPIC } from '../services/subscriptions'
import { withFilter } from 'graphql-subscriptions'
import moment from 'moment'

export const resolvers = {
  RootQuery: {
    async transactions (obj, { earliest }, { user }) {
      checkPermission(user)
      earliest = moment(earliest).toDate()

      return getTransactions(user, { earliest })
    }
  },

  RootSubscription: {
    transactionAdded: {
      subscribe: withFilter(() => pubsub.asyncIterator(TRANSACTION_ADDED_TOPIC), (payload, context) => {
        return payload.transactionAdded.user === context.user._id
      })
    }
  }
}
