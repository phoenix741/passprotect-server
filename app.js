'use strict';

import 'app-module-path/register';
import 'server/utils/promise';

import express from 'express';
import path from 'path';

import bootstrap from './server/bootstrap';
import middlewares from './server/middlewares';

const app = express();

bootstrap(app.get('env'));

// all environments
app.set('port', process.env.PORT || 3000);

// view engine setup
app.set('views', path.join(__dirname, 'common', 'templates'));
app.set('view engine', 'jade');

middlewares(app);
