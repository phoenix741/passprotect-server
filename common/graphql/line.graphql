# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

extend type RootQuery {
  # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]
}

union CreateUpdateLineResult = WalletLine | Errors

extend type RootMutation {
  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors
}
