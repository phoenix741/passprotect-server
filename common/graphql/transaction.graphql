# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

extend type RootQuery {
  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}
