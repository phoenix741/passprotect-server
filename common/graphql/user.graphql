# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

type RootQuery {
  # Request the list of users from the database.
  users: [User],

  # Request a user
  user(id: ID!): User,
}

union RegisterUserResult = User | Errors

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult
}
