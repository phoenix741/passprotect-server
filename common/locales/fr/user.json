{
  "login": {
    "dialog": {
      "title": "Connexion",
      "button": {
        "connect": "Continuer",
        "noaccount": "Vous n'avez pas encore de compte ? ",
        "signup": "Enregistrez-vous"
      }
    },
    "form": {
      "username": {
        "field": "Votre nom d'utilisateur",
        "help": {
          "cantbeblank": "Le nom de l'utilisateur ne peut être vide.",
          "invalid": "L'utilisateur ou le mot de passe sont incorrect.",
          "unknown": "Une erreur est survenue lors de l'authentification de l'utilisateur."
        }
      },
      "password": {
        "field": "Votre mot de passe",
        "help": {
          "default": "Vous avez oublié votre mot de passe ...",
          "cantbeblank": "Le mot de passe ne peut être vide."
        }
      }
    }
  },
  "register": {
    "form": {
      "title": "Nouveau compte",
      "identity_username": {
        "field": "Nom d'utilisateur"
      },
      "identity_password1": {
        "field": "Mot de passe"
      },
      "identity_password2": {
        "field": "Mot de passe (encore)"
      },
      "validation": {
        "field": "S'enregistrer"
      }
    },
    "flash": {
      "user_created": "Votre compte a bien été enregistré"
    }
  },
  "toolbar": {
    "notlogged": "Non connecté"
  }
}
