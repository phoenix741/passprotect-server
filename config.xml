<?xml version='1.0' encoding='utf-8'?>
<widget id="org.shadoware.passprotect" version="1.4.0" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>Passprotect</name>
    <description>
      This application is used to store your password and credit card in a secure store
    </description>
    <author email="ulrich.vdh@shadoware.org" href="https://passprotect.shadoware.org">
      The website
    </author>
    <content src="index.html" />
    <plugin name="cordova-plugin-whitelist" spec="1" />
    <plugin name="cordova-plugin-splashscreen" spec="^5.0.1" />
    <access origin="*" />
    <access origin="https://stats.shadoware.org" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <preference name="AutoHideSplashScreen" value="false" />
    <hook src="scripts/before_compile/build.js" type="before_compile" />
    <platform name="android">
        <allow-intent href="market:*" />
        <icon density="ldpi" src="res/android/icon/drawable-ldpi-icon.png" />
        <icon density="mdpi" src="res/android/icon/drawable-mdpi-icon.png" />
        <icon density="hdpi" src="res/android/icon/drawable-hdpi-icon.png" />
        <icon density="xhdpi" src="res/android/icon/drawable-xhdpi-icon.png" />
        <icon density="xxhdpi" src="res/android/icon/drawable-xxhdpi-icon.png" />
        <icon density="xxxhdpi" src="res/android/icon/drawable-xxxhdpi-icon.png" />
        <splash density="land-hdpi" src="res/android/splash/drawable-land-hdpi-screen.png" />
        <splash density="land-ldpi" src="res/android/splash/drawable-land-ldpi-screen.png" />
        <splash density="land-mdpi" src="res/android/splash/drawable-land-mdpi-screen.png" />
        <splash density="land-xhdpi" src="res/android/splash/drawable-land-xhdpi-screen.png" />
        <splash density="land-xxhdpi" src="res/android/splash/drawable-land-xxhdpi-screen.png" />
        <splash density="land-xxxhdpi" src="res/android/splash/drawable-land-xxxhdpi-screen.png" />
        <splash density="port-hdpi" src="res/android/splash/drawable-port-hdpi-screen.png" />
        <splash density="port-ldpi" src="res/android/splash/drawable-port-ldpi-screen.png" />
        <splash density="port-mdpi" src="res/android/splash/drawable-port-mdpi-screen.png" />
        <splash density="port-xhdpi" src="res/android/splash/drawable-port-xhdpi-screen.png" />
        <splash density="port-xxhdpi" src="res/android/splash/drawable-port-xxhdpi-screen.png" />
        <splash density="port-xxxhdpi" src="res/android/splash/drawable-port-xxxhdpi-screen.png" />
    </platform>
    <platform name="ios">
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
        <icon height="180" src="res/ios/icon/icon-60@3x.png" width="180" />
        <icon height="60" src="res/ios/icon/icon-60.png" width="60" />
        <icon height="120" src="res/ios/icon/icon-60@2x.png" width="120" />
        <icon height="76" src="res/ios/icon/icon-76.png" width="76" />
        <icon height="152" src="res/ios/icon/icon-76@2x.png" width="152" />
        <icon height="40" src="res/ios/icon/icon-40.png" width="40" />
        <icon height="80" src="res/ios/icon/icon-40@2x.png" width="80" />
        <icon height="120" src="res/ios/icon/icon-40@3x.png" width="120" />
        <icon height="57" src="res/ios/icon/icon.png" width="57" />
        <icon height="114" src="res/ios/icon/icon@2x.png" width="114" />
        <icon height="72" src="res/ios/icon/icon-72.png" width="72" />
        <icon height="144" src="res/ios/icon/icon-72@2x.png" width="144" />
        <icon height="29" src="res/ios/icon/icon-small.png" width="29" />
        <icon height="58" src="res/ios/icon/icon-small@2x.png" width="58" />
        <icon height="87" src="res/ios/icon/icon-small@3x.png" width="87" />
        <icon height="50" src="res/ios/icon/icon-50.png" width="50" />
        <icon height="100" src="res/ios/icon/icon-50@2x.png" width="100" />
        <icon height="167" src="res/ios/icon/icon-83.5@2x.png" width="167" />
        <splash height="480" src="res/screen/ios/Default~iphone.png" width="320" />
        <splash height="960" src="res/screen/ios/Default@2x~iphone.png" width="640" />
        <splash height="1024" src="res/screen/ios/Default-Portrait~ipad.png" width="768" />
        <splash height="2048" src="res/screen/ios/Default-Portrait@2x~ipad.png" width="1536" />
        <splash height="2732" src="res/screen/ios/Default-Portrait@~ipadpro.png" width="2048" />
        <splash height="768" src="res/screen/ios/Default-Landscape~ipad.png" width="1024" />
        <splash height="1536" src="res/screen/ios/Default-Landscape@2x~ipad.png" width="2048" />
        <splash height="2048" src="res/screen/ios/Default-Landscape@~ipadpro.png" width="2732" />
        <splash height="1136" src="res/screen/ios/Default-568h@2x~iphone.png" width="640" />
        <splash height="1334" src="res/screen/ios/Default-667h.png" width="750" />
        <splash height="2208" src="res/screen/ios/Default-736h.png" width="1242" />
        <splash height="1242" src="res/screen/ios/Default-Landscape-736h.png" width="2208" />
    </platform>
    <engine name="android" spec="7.0.0" />
</widget>
