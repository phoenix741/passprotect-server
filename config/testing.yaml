config:
    mongodb:
        host: 'mongodb://localhost:27017/passprotect'
        user: ~
        password: ~
        options: ~
