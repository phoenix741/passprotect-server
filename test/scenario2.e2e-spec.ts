import 'dotenv/config';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';
import {
  REGISTER_USER_QUERY,
  REGISTER_USER_1_VARIABLES,
  REGISTER_USER_2_VARIABLES,
  LOGIN_USER_QUERY,
  LOGIN_USER_1_VARIABLES,
  LOGIN_USER_2_VARIABLES,
} from './scenario2.data';
import { Model } from 'mongoose';
import { UserEntity } from '../src/users/models/user.entity';
import { getModelToken } from '@nestjs/mongoose';
import { LineEntity } from '../src/lines/models/line.entity';
import { TransactionEntity } from '../src/transactions/models/transaction.entity';

describe('AppController (e2e)', () => {
  let app;
  let userModel: Model<UserEntity>;
  let lineModel: Model<LineEntity>;
  let transactionModel: Model<TransactionEntity>;
  let connectionToken: string;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    userModel = moduleFixture.get<Model<UserEntity>>(getModelToken('User'));
    lineModel = moduleFixture.get<Model<LineEntity>>(getModelToken('Line'));
    transactionModel = moduleFixture.get<Model<TransactionEntity>>(
      getModelToken('Transaction'),
    );
  });

  function createRequest(query, variables) {
    return request(app.getHttpServer())
      .post('/graphql')
      .set('authorization', connectionToken || '')
      .set('content-type', 'application/json')
      .send({
        operationName: null,
        variables,
        query,
      });
  }

  it('Remove old user', async () => {
    await Promise.all(
      ['demo2', 'demo3'].map(async (user) => {
        await userModel.deleteMany({ _id: user });
        await lineModel.deleteMany({ user });
        await transactionModel.deleteMany({ user });
      }),
    );
  });

  it('Register user 1', () => {
    return createRequest(REGISTER_USER_QUERY, REGISTER_USER_1_VARIABLES)
      .expect(200)
      .expect(res => {
        expect(res.body.data.registerUser).toMatchSnapshot();
      });
  });

  it('Register user 2', () => {
    return createRequest(REGISTER_USER_QUERY, REGISTER_USER_2_VARIABLES)
      .expect(200)
      .expect(res => {
        expect(res.body.data.registerUser).toMatchSnapshot();
      });
  });

  it('Login the user 1', () => {
    return createRequest(LOGIN_USER_QUERY, LOGIN_USER_1_VARIABLES)
      .expect(200)
      .expect(res => {
        expect(res.body.data.createSession).toMatchSnapshot({
          token: expect.any(String),
        });
        connectionToken = res.body.data.createSession.token;
      });
  });

  it('Login the user 2', () => {
    return createRequest(LOGIN_USER_QUERY, LOGIN_USER_2_VARIABLES)
      .expect(200)
      .expect(res => {
        expect(res.body.data.createSession).toMatchSnapshot({
          token: expect.any(String),
        });
        connectionToken = res.body.data.createSession.token;
      });
  });

  afterAll(async () => {
    await app.close();
  });
});
