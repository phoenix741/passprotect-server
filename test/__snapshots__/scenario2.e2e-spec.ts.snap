// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) Login the user 1 1`] = `
Object {
  "__typename": "ConnectionInformation",
  "token": Any<String>,
  "user": Object {
    "_id": "demo2",
    "encryption": Object {
      "authTag": "dGVzdA==",
      "content": "dGVzdA==",
      "iv": "dGVzdA==",
      "salt": "dGVzdA==",
    },
  },
}
`;

exports[`AppController (e2e) Login the user 2 1`] = `
Object {
  "__typename": "ConnectionInformation",
  "token": Any<String>,
  "user": Object {
    "_id": "demo3",
    "encryption": Object {
      "authTag": "dGVzdA==",
      "content": "dGVzdA==",
      "iv": "dGVzdA==",
      "salt": "dGVzdA==",
    },
  },
}
`;

exports[`AppController (e2e) Register user 1 1`] = `
Object {
  "__typename": "User",
  "_id": "demo2",
}
`;

exports[`AppController (e2e) Register user 2 1`] = `
Object {
  "__typename": "User",
  "_id": "demo3",
}
`;
